File: server/logger.ts
import { appendFile, readFile, mkdir, readdir, stat } from "fs/promises";
import path from "path";
import { createHash } from "crypto";

const LOGS_DIR = path.resolve("logs");
const TRANSCRIPTS_DIR = path.join(LOGS_DIR, "transcripts");
const AI_TRANSCRIPTS_DIR = path.join(LOGS_DIR, "ai-transcripts");

const LOG_STREAMS = {
  master: "a0p-master.jsonl",
  edcm: "edcm-metrics.jsonl",
  memory: "memory-tensor.jsonl",
  sentinel: "sentinel-memory.jsonl",
  interference: "memory-interference.jsonl",
  attribution: "memory-attribution.jsonl",
  omega: "omega-autonomy.jsonl",
} as const;

export type LogStream = keyof typeof LOG_STREAMS;

interface LogEntry {
  timestamp: string;
  stream: string;
  subsystem: string;
  event: string;
  data: Record<string, any>;
}

let loggingEnabled = true;
let streamToggles: Record<string, boolean> = {
  master: true,
  edcm: true,
  memory: true,
  sentinel: true,
  interference: true,
  attribution: true,
  omega: true,
  transcripts: true,
  "ai-transcripts": true,
};

let initialized = false;

async function ensureDirs() {
  if (initialized) return;
  await mkdir(LOGS_DIR, { recursive: true });
  await mkdir(TRANSCRIPTS_DIR, { recursive: true });
  await mkdir(AI_TRANSCRIPTS_DIR, { recursive: true });
  initialized = true;
}

function buildEntry(stream: string, subsystem: string, event: string, data: Record<string, any>): LogEntry {
  return {
    timestamp: new Date().toISOString(),
    stream,
    subsystem,
    event,
    data,
  };
}

async function appendToFile(filePath: string, entry: LogEntry): Promise<void> {
  await ensureDirs();
  await appendFile(filePath, JSON.stringify(entry) + "\n", "utf-8");
}

function getStreamPath(stream: LogStream): string {
  return path.join(LOGS_DIR, LOG_STREAMS[stream]);
}

export function setLoggingEnabled(enabled: boolean): void {
  loggingEnabled = enabled;
}

export function setStreamToggle(stream: string, enabled: boolean): void {
  streamToggles[stream] = enabled;
}

export function getStreamToggles(): Record<string, boolean> {
  return { ...streamToggles };
}

export function isStreamEnabled(stream: string): boolean {
  return loggingEnabled && (streamToggles[stream] !== false);
}

export function updateTogglesFromSystem(params: Record<string, any> | null | undefined): void {
  if (!params) return;
  if (typeof params.enabled === "boolean") {
    loggingEnabled = params.enabled;
  }
  if (params.streams && typeof params.streams === "object") {
    for (const [k, v] of Object.entries(params.streams)) {
      if (typeof v === "boolean") {
        streamToggles[k] = v;
      }
    }
  }
}

export async function logMaster(subsystem: string, event: string, data: Record<string, any>): Promise<void> {
  if (!isStreamEnabled("master")) return;
  const entry = buildEntry("master", subsystem, event, data);
  try {
    await appendToFile(getStreamPath("master"), entry);
  } catch (e) {
    console.error("Logger master write error:", e);
  }
}

export async function logEdcm(event: string, data: Record<string, any>): Promise<void> {
  if (!isStreamEnabled("edcm")) return;
  const entry = buildEntry("edcm", "edcm", event, data);
  try {
    await appendToFile(getStreamPath("edcm"), entry);
    if (isStreamEnabled("master")) {
      await appendToFile(getStreamPath("master"), entry);
    }
  } catch (e) {
    console.error("Logger edcm write error:", e);
  }
}

export async function logMemory(event: string, data: Record<string, any>): Promise<void> {
  if (!isStreamEnabled("memory")) return;
  const entry = buildEntry("memory", "memory_tensor", event, data);
  try {
    await appendToFile(getStreamPath("memory"), entry);
    if (isStreamEnabled("master")) {
      await appendToFile(getStreamPath("master"), entry);
    }
  } catch (e) {
    console.error("Logger memory write error:", e);
  }
}

export async function logSentinel(event: string, data: Record<string, any>): Promise<void> {
  if (!isStreamEnabled("sentinel")) return;
  const entry = buildEntry("sentinel", "sentinel", event, data);
  try {
    await appendToFile(getStreamPath("sentinel"), entry);
    if (isStreamEnabled("master")) {
      await appendToFile(getStreamPath("master"), entry);
    }
  } catch (e) {
    console.error("Logger sentinel write error:", e);
  }
}

export async function logInterference(event: string, data: Record<string, any>): Promise<void> {
  if (!isStreamEnabled("interference")) return;
  const entry = buildEntry("interference", "memory_interference", event, data);
  try {
    await appendToFile(getStreamPath("interference"), entry);
    if (isStreamEnabled("master")) {
      await appendToFile(getStreamPath("master"), entry);
    }
  } catch (e) {
    console.error("Logger interference write error:", e);
  }
}

export async function logAttribution(event: string, data: Record<string, any>): Promise<void> {
  if (!isStreamEnabled("attribution")) return;
  const entry = buildEntry("attribution", "memory_attribution", event, data);
  try {
    await appendToFile(getStreamPath("attribution"), entry);
    if (isStreamEnabled("master")) {
      await appendToFile(getStreamPath("master"), entry);
    }
  } catch (e) {
    console.error("Logger attribution write error:", e);
  }
}

export async function logOmega(event: string, data: Record<string, any>): Promise<void> {
  if (!isStreamEnabled("omega")) return;
  const entry = buildEntry("omega", "omega_autonomy", event, data);
  try {
    await appendToFile(getStreamPath("omega"), entry);
    if (isStreamEnabled("master")) {
      await appendToFile(getStreamPath("master"), entry);
    }
  } catch (e) {
    console.error("Logger omega write error:", e);
  }
}

export async function logTranscript(transcriptHash: string, event: string, data: Record<string, any>): Promise<void> {
  if (!isStreamEnabled("transcripts")) return;
  const timestamp = Date.now();
  const hash = transcriptHash || createHash("sha256").update(String(timestamp)).digest("hex").slice(0, 12);
  const filename = `transcript-${timestamp}-${hash}.jsonl`;
  const filePath = path.join(TRANSCRIPTS_DIR, filename);
  const entry = buildEntry("transcript", "transcript", event, { ...data, transcriptHash: hash });
  try {
    await appendToFile(filePath, entry);
    if (isStreamEnabled("master")) {
      await appendToFile(getStreamPath("master"), { ...entry, data: { ...entry.data, transcriptFile: filename } });
    }
  } catch (e) {
    console.error("Logger transcript write error:", e);
  }
  return;
}

export async function appendToTranscript(filename: string, event: string, data: Record<string, any>): Promise<void> {
  if (!isStreamEnabled("transcripts")) return;
  const filePath = path.join(TRANSCRIPTS_DIR, filename);
  const entry = buildEntry("transcript", "transcript", event, data);
  try {
    await appendToFile(filePath, entry);
  } catch (e) {
    console.error("Logger transcript append error:", e);
  }
}

export async function readLogStream(stream: LogStream, options: { offset?: number; limit?: number } = {}): Promise<{ entries: LogEntry[]; total: number }> {
  const filePath = getStreamPath(stream);
  try {
    const content = await readFile(filePath, "utf-8");
    const lines = content.trim().split("\n").filter(Boolean);
    const total = lines.length;
    const offset = options.offset || 0;
    const limit = options.limit || 100;
    const startIdx = Math.max(0, total - offset - limit);
    const endIdx = total - offset;
    const slice = lines.slice(Math.max(0, startIdx), Math.max(0, endIdx));
    const entries = slice.map((line) => {
      try {
        return JSON.parse(line);
      } catch {
        return { raw: line };
      }
    }).reverse();
    return { entries, total };
  } catch (e: any) {
    if (e.code === "ENOENT") return { entries: [], total: 0 };
    throw e;
  }
}

export async function readTranscriptLog(filename: string): Promise<LogEntry[]> {
  const filePath = path.join(TRANSCRIPTS_DIR, filename);
  try {
    const content = await readFile(filePath, "utf-8");
    return content.trim().split("\n").filter(Boolean).map((line) => {
      try {
        return JSON.parse(line);
      } catch {
        return { raw: line } as any;
      }
    });
  } catch (e: any) {
    if (e.code === "ENOENT") return [];
    throw e;
  }
}

export async function listTranscripts(): Promise<{ filename: string; size: number; created: string }[]> {
  await ensureDirs();
  try {
    const files = await readdir(TRANSCRIPTS_DIR);
    const transcriptFiles = files.filter((f) => f.endsWith(".jsonl")).sort().reverse();
    const results = [];
    for (const f of transcriptFiles) {
      const filePath = path.join(TRANSCRIPTS_DIR, f);
      const s = await stat(filePath);
      results.push({ filename: f, size: s.size, created: s.birthtime.toISOString() });
    }
    return results;
  } catch {
    return [];
  }
}

export async function getLogStats(): Promise<Record<string, { size: number; lines: number }>> {
  await ensureDirs();
  const stats: Record<string, { size: number; lines: number }> = {};
  for (const [stream, filename] of Object.entries(LOG_STREAMS)) {
    const filePath = path.join(LOGS_DIR, filename);
    try {
      const s = await stat(filePath);
      const content = await readFile(filePath, "utf-8");
      const lines = content.trim().split("\n").filter(Boolean).length;
      stats[stream] = { size: s.size, lines };
    } catch {
      stats[stream] = { size: 0, lines: 0 };
    }
  }
  return stats;
}

export interface AiTranscriptEntry {
  timestamp: string;
  conversationId?: number;
  messageId?: number;
  model: string;
  request: any;
  response: string;
  tokens: { prompt: number; completion: number; total: number };
  latencyMs: number;
  status: "success" | "error";
  error?: string;
}

function getAiTranscriptFilePath(): string {
  const now = new Date();
  const dateStr = now.toISOString().slice(0, 10);
  return path.join(AI_TRANSCRIPTS_DIR, `ai-transcript-${dateStr}.jsonl`);
}

export async function logAiTranscript(entry: AiTranscriptEntry): Promise<void> {
  if (!isStreamEnabled("ai-transcripts")) return;
  await ensureDirs();
  const filePath = getAiTranscriptFilePath();
  try {
    await appendFile(filePath, JSON.stringify(entry) + "\n", "utf-8");
  } catch (e) {
    console.error("Logger ai-transcript write error:", e);
  }
}

export async function readAiTranscripts(options: {
  date?: string;
  model?: string;
  offset?: number;
  limit?: number;
} = {}): Promise<{ entries: AiTranscriptEntry[]; total: number }> {
  await ensureDirs();
  try {
    let files: string[];
    if (options.date) {
      files = [`ai-transcript-${options.date}.jsonl`];
    } else {
      const allFiles = await readdir(AI_TRANSCRIPTS_DIR);
      files = allFiles.filter(f => f.startsWith("ai-transcript-") && f.endsWith(".jsonl")).sort().reverse();
    }

    let allEntries: AiTranscriptEntry[] = [];
    for (const file of files) {
      const filePath = path.join(AI_TRANSCRIPTS_DIR, file);
      try {
        const content = await readFile(filePath, "utf-8");
        const lines = content.trim().split("\n").filter(Boolean);
        for (const line of lines) {
          try {
            const entry = JSON.parse(line) as AiTranscriptEntry;
            if (options.model && entry.model !== options.model) continue;
            allEntries.push(entry);
          } catch {}
        }
      } catch (e: any) {
        if (e.code !== "ENOENT") throw e;
      }
    }

    allEntries.sort((a, b) => new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime());
    const total = allEntries.length;
    const offset = options.offset || 0;
    const limit = options.limit || 50;
    const slice = allEntries.slice(offset, offset + limit);
    return { entries: slice, total };
  } catch (e: any) {
    if (e.code === "ENOENT") return { entries: [], total: 0 };
    throw e;
  }
}

export async function listAiTranscriptFiles(): Promise<{ filename: string; size: number; date: string }[]> {
  await ensureDirs();
  try {
    const files = await readdir(AI_TRANSCRIPTS_DIR);
    const transcriptFiles = files.filter(f => f.startsWith("ai-transcript-") && f.endsWith(".jsonl")).sort().reverse();
    const results = [];
    for (const f of transcriptFiles) {
      const filePath = path.join(AI_TRANSCRIPTS_DIR, f);
      const s = await stat(filePath);
      const dateMatch = f.match(/ai-transcript-(\d{4}-\d{2}-\d{2})\.jsonl/);
      results.push({ filename: f, size: s.size, date: dateMatch ? dateMatch[1] : "" });
    }
    return results;
  } catch {
    return [];
  }
}

File: python/logger.py
# 268:4
# NOTE: 447 lines — over the 400-line guideline. Split on next modification.
import asyncio
import json
import os
from datetime import datetime
from hashlib import sha256
from pathlib import Path
from typing import Any, Dict, List, Optional

LOGS_DIR = Path("logs").resolve()
TRANSCRIPTS_DIR = LOGS_DIR / "transcripts"
AI_TRANSCRIPTS_DIR = LOGS_DIR / "ai-transcripts"

LOG_STREAMS: Dict[str, str] = {
    "master": "a0p-master.jsonl",
    "edcm": "edcm-metrics.jsonl",
    "memory": "memory-tensor.jsonl",
    "sentinel": "sentinel-memory.jsonl",
    "interference": "memory-interference.jsonl",
    "attribution": "memory-attribution.jsonl",
    "omega": "omega-autonomy.jsonl",
    "psi": "psi-selfmodel.jsonl",
    "openai_events": "openai-events.jsonl",
    "openai_hmmm": "openai-hmmm.jsonl",
}

logging_enabled: bool = True
stream_toggles: Dict[str, bool] = {
    "master": True,
    "edcm": True,
    "memory": True,
    "sentinel": True,
    "interference": True,
    "attribution": True,
    "omega": True,
    "psi": True,
    "transcripts": True,
    "ai-transcripts": True,
    "openai_events": True,
    "openai_hmmm": True,
}
_initialized: bool = False


async def _ensure_dirs() -> None:
    global _initialized
    if _initialized:
        return
    LOGS_DIR.mkdir(parents=True, exist_ok=True)
    TRANSCRIPTS_DIR.mkdir(parents=True, exist_ok=True)
    AI_TRANSCRIPTS_DIR.mkdir(parents=True, exist_ok=True)
    _initialized = True


def _build_entry(stream: str, subsystem: str, event: str, data: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "timestamp": datetime.utcnow().isoformat() + "Z",
        "stream": stream,
        "subsystem": subsystem,
        "event": event,
        "data": data,
    }


async def _append_to_file(file_path: Path, entry: Dict[str, Any]) -> None:
    await _ensure_dirs()
    line = json.dumps(entry) + "\n"
    file_path.parent.mkdir(parents=True, exist_ok=True)
    with open(file_path, "a", encoding="utf-8") as f:
        f.write(line)


def _get_stream_path(stream: str) -> Path:
    return LOGS_DIR / LOG_STREAMS[stream]


def set_logging_enabled(enabled: bool) -> None:
    global logging_enabled
    logging_enabled = enabled


def set_stream_toggle(stream: str, enabled: bool) -> None:
    stream_toggles[stream] = enabled


def get_stream_toggles() -> Dict[str, bool]:
    return dict(stream_toggles)


def is_stream_enabled(stream: str) -> bool:
    return logging_enabled and stream_toggles.get(stream, True)


def update_toggles_from_system(params: Optional[Dict[str, Any]]) -> None:
    global logging_enabled
    if not params:
        return
    if isinstance(params.get("enabled"), bool):
        logging_enabled = params["enabled"]
    if isinstance(params.get("streams"), dict):
        for k, v in params["streams"].items():
            if isinstance(v, bool):
                stream_toggles[k] = v


async def log_master(subsystem: str, event: str, data: Dict[str, Any]) -> None:
    if not is_stream_enabled("master"):
        return
    entry = _build_entry("master", subsystem, event, data)
    try:
        await _append_to_file(_get_stream_path("master"), entry)
    except Exception as e:
        print(f"Logger master write error: {e}")


async def log_edcm(event: str, data: Dict[str, Any]) -> None:
    if not is_stream_enabled("edcm"):
        return
    entry = _build_entry("edcm", "edcm", event, data)
    try:
        await _append_to_file(_get_stream_path("edcm"), entry)
        if is_stream_enabled("master"):
            await _append_to_file(_get_stream_path("master"), entry)
    except Exception as e:
        print(f"Logger edcm write error: {e}")


async def log_memory(event: str, data: Dict[str, Any]) -> None:
    if not is_stream_enabled("memory"):
        return
    entry = _build_entry("memory", "memory_tensor", event, data)
    try:
        await _append_to_file(_get_stream_path("memory"), entry)
        if is_stream_enabled("master"):
            await _append_to_file(_get_stream_path("master"), entry)
    except Exception as e:
        print(f"Logger memory write error: {e}")


async def log_sentinel(event: str, data: Dict[str, Any]) -> None:
    if not is_stream_enabled("sentinel"):
        return
    entry = _build_entry("sentinel", "sentinel", event, data)
    try:
        await _append_to_file(_get_stream_path("sentinel"), entry)
        if is_stream_enabled("master"):
            await _append_to_file(_get_stream_path("master"), entry)
    except Exception as e:
        print(f"Logger sentinel write error: {e}")


async def log_interference(event: str, data: Dict[str, Any]) -> None:
    if not is_stream_enabled("interference"):
        return
    entry = _build_entry("interference", "memory_interference", event, data)
    try:
        await _append_to_file(_get_stream_path("interference"), entry)
        if is_stream_enabled("master"):
            await _append_to_file(_get_stream_path("master"), entry)
    except Exception as e:
        print(f"Logger interference write error: {e}")


async def log_attribution(event: str, data: Dict[str, Any]) -> None:
    if not is_stream_enabled("attribution"):
        return
    entry = _build_entry("attribution", "memory_attribution", event, data)
    try:
        await _append_to_file(_get_stream_path("attribution"), entry)
        if is_stream_enabled("master"):
            await _append_to_file(_get_stream_path("master"), entry)
    except Exception as e:
        print(f"Logger attribution write error: {e}")


async def log_omega(event: str, data: Dict[str, Any]) -> None:
    if not is_stream_enabled("omega"):
        return
    entry = _build_entry("omega", "omega_autonomy", event, data)
    try:
        await _append_to_file(_get_stream_path("omega"), entry)
        if is_stream_enabled("master"):
            await _append_to_file(_get_stream_path("master"), entry)
    except Exception as e:
        print(f"Logger omega write error: {e}")


async def log_psi(event: str, data: Dict[str, Any]) -> None:
    if not is_stream_enabled("psi"):
        return
    entry = _build_entry("psi", "psi_selfmodel", event, data)
    try:
        await _append_to_file(_get_stream_path("psi"), entry)
        if is_stream_enabled("master"):
            await _append_to_file(_get_stream_path("master"), entry)
    except Exception as e:
        print(f"Logger psi write error: {e}")


async def log_transcript(transcript_hash: str, event: str, data: Dict[str, Any]) -> None:
    if not is_stream_enabled("transcripts"):
        return
    timestamp = int(datetime.utcnow().timestamp() * 1000)
    h = transcript_hash or sha256(str(timestamp).encode()).hexdigest()[:12]
    filename = f"transcript-{timestamp}-{h}.jsonl"
    file_path = TRANSCRIPTS_DIR / filename
    entry = _build_entry("transcript", "transcript", event, {**data, "transcriptHash": h})
    try:
        await _append_to_file(file_path, entry)
        if is_stream_enabled("master"):
            await _append_to_file(_get_stream_path("master"), {**entry, "data": {**entry["data"], "transcriptFile": filename}})
    except Exception as e:
        print(f"Logger transcript write error: {e}")


async def append_to_transcript(filename: str, event: str, data: Dict[str, Any]) -> None:
    if not is_stream_enabled("transcripts"):
        return
    file_path = TRANSCRIPTS_DIR / filename
    entry = _build_entry("transcript", "transcript", event, data)
    try:
        await _append_to_file(file_path, entry)
    except Exception as e:
        print(f"Logger transcript append error: {e}")


async def read_log_stream(stream: str, offset: int = 0, limit: int = 100) -> Dict[str, Any]:
    file_path = _get_stream_path(stream)
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()
        lines = [l for l in content.strip().split("\n") if l]
        total = len(lines)
        start_idx = max(0, total - offset - limit)
        end_idx = max(0, total - offset)
        sliced = lines[start_idx:end_idx]
        entries = []
        for line in reversed(sliced):
            try:
                entries.append(json.loads(line))
            except Exception:
                entries.append({"raw": line})
        return {"entries": entries, "total": total}
    except FileNotFoundError:
        return {"entries": [], "total": 0}


async def read_transcript_log(filename: str) -> List[Dict[str, Any]]:
    file_path = TRANSCRIPTS_DIR / filename
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()
        result = []
        for line in content.strip().split("\n"):
            if not line:
                continue
            try:
                result.append(json.loads(line))
            except Exception:
                result.append({"raw": line})
        return result
    except FileNotFoundError:
        return []


async def list_transcripts() -> List[Dict[str, Any]]:
    await _ensure_dirs()
    try:
        files = sorted(
            [f for f in TRANSCRIPTS_DIR.iterdir() if f.suffix == ".jsonl"],
            key=lambda x: x.name,
            reverse=True,
        )
        results = []
        for f in files:
            stat = f.stat()
            results.append({
                "filename": f.name,
                "size": stat.st_size,
                "created": datetime.fromtimestamp(stat.st_ctime).isoformat(),
            })
        return results
    except Exception:
        return []


async def get_log_stats() -> Dict[str, Dict[str, Any]]:
    await _ensure_dirs()
    stats: Dict[str, Dict[str, Any]] = {}
    for stream, filename in LOG_STREAMS.items():
        file_path = LOGS_DIR / filename
        try:
            stat = file_path.stat()
            with open(file_path, "r", encoding="utf-8") as f:
                content = f.read()
            lines = len([l for l in content.strip().split("\n") if l])
            stats[stream] = {"size": stat.st_size, "lines": lines}
        except FileNotFoundError:
            stats[stream] = {"size": 0, "lines": 0}
    return stats


# AI-transcript and OpenAI-event helpers have moved to logger_ai.py.
# Re-exported here for backward-compat (callers that do
# `from .logger import log_openai_event` continue to work).
from .logger_ai import (  # noqa: E402
    log_ai_transcript,
    read_ai_transcripts,
    log_openai_event,
    seed_openai_hmmm_if_empty,
    append_openai_hmmm,
    read_openai_hmmm,
    list_ai_transcript_files,
)

__all__ = [
    "log_ai_transcript",
    "read_ai_transcripts",
    "log_openai_event",
    "seed_openai_hmmm_if_empty",
    "append_openai_hmmm",
    "read_openai_hmmm",
    "list_ai_transcript_files",
]
# 268:4
